README.md
# X-STILT: Column-Stochastic Time-Inverted Lagrangian Transport model

X-STILT is an atmospheric transport model that deals with vertically integrated column concentrations of various trace gases ([Wu et al., 2018, 2023](https://doi.org/10.5194/gmd-11-4843-2018)). The model code was built upon the Stochastic Time-Inverted Lagrangian Transport (STILT) model ([Lin et al., 2003](https://doi.org/10.1029/2002JD003161)) and its [latest version 2](https://github.com/uataq/stilt) ([Fasoli et al., 2018](https://doi.org/10.5194/gmd-11-2813-2018)). 

The model framework can now work with OCO-2&3 XCO<sub>2</sub> and TROPOMI column CO ([Wu et al., ACP](https://acp.copernicus.org/articles/22/14547/2022/acp-22-14547-2022.html)), CH<sub>4</sub> (Li et al., and Tribby et al., in prep), and NO<sub>2</sub> ([Wu et al., EGUsphere: STILT-NOx](https://egusphere.copernicus.org/preprints/2023/egusphere-2023-876/)) and is testing for [TCCON and EM27](https://github.com/uataq/X-STILT/pull/6). :sunglasses: 

This GitHub repo includes built-in scripts/functions for 
1. running backward trajectories from an atmospheric column and column footprint (start with `run_xstilt.r` for either initial X-STILT or updated STILT-NOx); 
2. running forward-time trajectories from a box around the site and compute background from satellite observations (start with `compute_bg.r`); 
3. estimating wind and PBL uncertainties and translating those into XCO<sub>2</sub> uncertainties (start with `run_xstilt.r`)
4. simulating column CO<sub>2</sub>, CO, and NO<sub>2</sub> abundances at OCO-2/3 and TROPOMI soundings (STILT-NOx, start with `run_sim_nox.r`)

Model developments are ongoing and contributions are welcomed and appreciated. Please contact Dien (dienwu@caltech.edu) if you are interested in other column sensors/species or have any questions.

# X-STILT Features
## Table of Contents
- [**Recent commits**](HISTORY.md)
- [**Download and install model**](#download-and-install-model)
- [**Prerequisites**](#prerequisites)
- [**Obtain column footprint**](#obtain-column-footprint)
- [**Determine background XCO<sub>2</sub>**](#determine-background-xco2)
- [**Estimate horizontal and vertical transport errors**](#estimate-horizontal-and-vertical-transport-errors)
- [**Atmospheric inversion on XCO<sub>2</sub>**](#atmospheric-inversion-on-xco2)
- [**Example figures of column footprints and XCO<sub>2.ff</sub>**](#example-figures-of-column-footprints-and-xco2ff)
- [**Reference**](#reference)

Download and install model
============
   * As STILT-R version 2 now serves as a submodule of X-STILT, you will need to use the git command to download the entire model packages including the underlying STILTv2 and HYSPLITv5:
   ```
   git clone --recursive https://github.com/uataq/X-STILT.git
   ```

   * To automatically install all R packages (requiring R > 3.5.0) and the STILTv2 model with the fortran program for HYSPLITv5, start with the [first few lines of `run_xstilt.r`](https://github.com/uataq/X-STILT/blob/master/run_xstilt.r#L45-L47). 

   * Required datasets and methodologies are described in [prerequisites](prerequisites) and [*Wu et al*., 2018](https://doi.org/10.5194/gmd-11-4843-2018). Please refer to the [reference](#reference) section if you plan to publish any study using the (X-)STILT model. 


<!-- toc -->
Prerequisites
============
1. For **SATELLITE-dependent column simulation** - 
   * download [OCO-2 Lite](https://disc.gsfc.nasa.gov/datasets?keywords=OCO%20L2%20Lite%20FP&page=1) or [TROPOMI](https://disc.gsfc.nasa.gov/datasets?keywords=TROPOMI%20L2&page=1) Level 2 files and modify the corresponding parameters including `obs_sensor`, `obs_ver`, `obs_species`, `oco_path`, or `trp_path`. 
   * X-STILT reads in averaging kernels from satellite files but calculates its own pressure weighting functions to perform vertical weighting on footprint values for air parcels that releases from different altitudes and then provide the vertically compressed column footprints. 

2. For **IDEAL column simulation** WITHOUT satellite dependence - 
   * prepare a csv file for receptor `lon`, `lat`, and `time`. An example is provided as in `receptor_demo.csv`. Check out instructions in [recent commit](https://github.com/uataq/X-STILT/blob/master/HISTORY.md#commit-by-july-6-2021). 


3. Meteorological fields in ARL format, e.g., ones provided by [NOAA ARL](ftp://arlftp.arlhq.noaa.gov/archives/), and modify the section on [ARL format meteo params](https://github.com/uataq/X-STILT/blob/master/run_xstilt.r#L164-L179). One need to perform ARL conversion, e.g., when using WRF as the met fields. 

4. (OPTIONAL) For calculating column enhancements [in ppm] - 
   * choose your favourite emission inventories. The default is to use ODIAC for fossil fuel emission representation; so download [1 km ODIAC files](http://db.cger.nies.go.jp/dataset/ODIAC/DL_odiac2019.html) in tif format and modify `odiac_path`. 

5. (OPTIONAL) For performing transport error analyses - 
   * [NOAA radiosonde data](https://ruc.noaa.gov/raobs/) for computing model-data wind errors; please choose wind speed unit of tenths of m s-1 and FSL format; Users could download RAOB stations within a spatial area around `site`;
   * [Carbon-Tracker mole fraction data, e.g., CT-NRT](https://www.esrl.noaa.gov/gmd/ccgg/carbontracker/CT-NRT/) for getting the total XCO<sub>2</sub> in addition to FF XCO<sub>2</sub>. 

6. (OPTIONAL) For performing emission error analyses - 
   * Bottom-up emission inventories ensemble [FFDAS](http://ffdas.rc.nau.edu/index.html) and [EDGAR](https://edgar.jrc.ec.europa.eu/). 


Obtain column footprint
============
*Column Footprints* [ppm / (umol m<sup>-2</sup> s<sup>-1</sup>)] are the source-receptor sensivities or essentially the Jacobian Matrics between concentration (enhancements) and fluxes (for a given source/sink). Users can start with `run_xstilt.r` for model and parameter initializations.

1. Select a satellite overpass. Either by a manual insertion Or an automatic search. By default, [STEP 1](https://github.com/uataq/X-STILT/blob/master/run_xstilt.r#L102-L116) searches for all overpasses that have soundings falling into a spatial domain (e.g., [2 deg x 2 deg](https://github.com/uataq/X-STILT/blob/master/run_xstilt.r#L57-L58)) as well as a near-field domain (i.e., [0.6 deg x 0.6 deg](https://github.com/uataq/X-STILT/blob/master/run_xstilt.r#L59-L60)).

2. Select the kind of simulation one would like to conduct in [STEP 2](https://github.com/uataq/X-STILT/blob/master/run_xstilt.r#L120-L132). One needs to modify logical flags which included:

   >* `run_trajec = T or run_foot = T`: Backward trajectories + vertically weighted column footprints;

   >* `run_hor_err = T`: Horizontal transport error analysis (one needs to calculate wind error statistics first, see below); 

   >* `run_wind_err = T`: Modeled wind error estimates against radiosonde data (one need to download them from NOAA webpage first); 

   >* `run_ver_err = T`: Vertical transport error analysis (via scaling mixed layer height up and down with scaling factors stated in [STEP 4](https://github.com/uataq/X-STILT/blob/master/run_xstilt.r#L253-L256));

   >* `run_sim = T`: Simulate FFCO2 XCO<sub>2</sub> enhancements (requires footprint), remember to turn off `run_trajec` and `run_foot`;

   >* `run_emiss_err = T`: Prior emission error analysis (requires footprint).

3. Modify parameters for placing column receptors (e.g., max height and vertical spacing between two vertical levels in meters, # of particles per level, receptor locations). By default, X-STILT chooses more receptors within NEAR-FIELD area, see [STEP 3](https://github.com/uataq/X-STILT/blob/master/run_xstilt.r#L135-L160).

4. Modify meteorological fields, [STEP 4](https://github.com/uataq/X-STILT/blob/master/run_xstilt.r#L164-L179). See [STILT documentation](https://uataq.github.io/stilt/#/configuration?id=meteorological-data-input) for more explanations. 

5. Modify horizontal resolution and spatial extent of footprint, [STEP 5](https://github.com/uataq/X-STILT/blob/master/run_xstilt.r#L182-L214). 
   * X-STILT can generate multiple sets of footprints based on the same set of trajectory in one simulation (hourly-integrated or hourly-explicit footprints). For generating a second or third set of footprints, modify these [params](https://github.com/uataq/X-STILT/blob/master/run_xstilt.r#L202-L214). 

6. Set up SLURM parallel computing, [STEP 6](https://github.com/uataq/X-STILT/blob/master/run_xstilt.r#L241-L260).


Determine background XCO<sub>2</sub>
============
Start with main script of `compute_bg.r`. [M3 is the overpass specific background [ppm]](https://github.com/uataq/X-STILT/blob/master/compute_bg.r) by releasing air parcels in a forward fashion from a city and determining urban plume and background region using 2D kernel density. Please refer to details described in [Sect. 2.3 in Wu et al. (2018)](https://www.geosci-model-dev.net/11/4843/2018/#section2).

   >* `run_trajec = T`: calculate forward-time trajectories from a small box around the site (see parameter `box.len` and other parameters in STEP 2) no matter if those trajectories had been calculated previously. In this mode, air parcels are released continuously for every `dtime.sep` hours (see parameter `dtime.*`);

   >* `run_trajec = T & run_for_err = T`: add a wind error component (controlled by parameter `siguverr`) while calculating forward-time trajectories;

   >* `run_wind_err = T`: calculate modeled wind errors against radiosonde data (one need to download them first);

   >* `run_bg = T`: calculate the background values based on OCO-2/3 or TROPOMI data (see parameters in STEP 4); the inner functions will generate plots of plumes and observed enhancements by assuming background region to the north/south/east/west outside the plume. The final estimated background values will be stored in `file.path(store_path, fn)` if `writeTF = T`. 


Estimate horizontal and vertical transport errors
============
Instead of using model ensembles for estimating errors, X-STILT propagates real-world random u-v- wind errors (with correlation length scale and timescale) into errors in XCO<sub>2</sub>. The fundamental approach is proposed and documented in [Lin and Gerbig, 2005](https://agupubs.onlinelibrary.wiley.com/doi/epdf/10.1029/2004GL021127), which is now extended to column CO2. We calculate and propagate XCO<sub>2</sub> errors from release levels, to receptor locations, and finally to overpasses. 

Briefly speaking, users need to first generate another set of trajectories with wind error statistics (i.e., blue dots in Fig. S4) by turning on `run_hor_err`. Once trajectories with wind errors have been generated, one need to turn on `run_sim` to let X-STILT calculate the errors in XCO<sub>2</sub> based on two sets of trajectories (with/without wind errors). Hard to explain all the steps here, but feel free to contact Dien if one is interested in performing such error estimates. Please also refer to details described in [Sect. 2.6 in Wu et al. (2018)](https://www.geosci-model-dev.net/11/4843/2018/#section2).

<img src="figs/trans_err.png" width="450" height="300">


Atmospheric inversion
============
As discussed in [Sect. 4.2 in Wu et al. (2018)](https://www.geosci-model-dev.net/11/4843/2018/#section4), we could come up the posterior scaling factor for anthropogenic emissions based on 5 overpasses over Riyadh, via a simple Bayesian Inversion. We treated the entire city as a whole and solve for one scaling factor, given biases in near-field wind direction. So, we did not solve for posterior emissions for every grid cell within a city. Codes are not included in this repo. 

Example figures of column footprints and XCO<sub>2.ff</sub>
============
<img src="figs/wgt_sum_xfoot_Riyadh_2015121610_gdas0p5_STILTv2_zoom8_-72hrs_100dpar.png" width="400" height="400">

Figure - Latitude integrated map of weighted column footprints [umol m-2 s-1] on 12/29/2014 from 70+ selected sounding/receptor over Riyadh.

<img src="figs/wgt_sum_xco2_Riyadh_2015121610_gdas0p5_STILTv2_zoom8_-72hrs_100dpar.png" width="400" height="400">

Figure - Latitude integrated XCO<sub>2</sub>.ff contribution maps [ppm] on 12/29/2014 from 70+ selected sounding/receptor over Riyadh.


Reference
============
## Model development 
**Initial X-STILTv1 paper**: 
Wu, D., Lin, J. C., Fasoli, B., Oda, T., Ye, X., Lauvaux, T., Yang, E. G., and Kort, E. A.: A Lagrangian approach towards extracting signals of urban CO2 emissions from satellite observations of atmospheric column CO2 (XCO<sub>2</sub>): X-Stochastic Time-Inverted Lagrangian Transport model (“X-STILT v1”), *Geosci. Model Dev.*, 11, 4843-4871, https://doi.org/10.5194/gmd-11-4843-2018, 2018. 

**STILTv2 paper**: 
Fasoli, B., Lin, J. C., Bowling, D. R., Mitchell, L., and Mendoza, D.: Simulating atmospheric tracer concentrations for spatially distributed receptors: updates to the Stochastic Time-Inverted Lagrangian Transport model's R interface (STILT-R version 2), *Geosci. Model Dev.*, 11, 2813-2824, https://doi.org/10.5194/gmd-11-2813-2018, 2018.

**Initial STILTv1 paper**: 
Lin, J.C., Gerbig, C., Wofsy, S.C., Andrews, A.E., Daube, B.C., Davis, K.J. and Grainger, C.A.: A near‐field tool for simulating the upstream influence of atmospheric observations: The Stochastic Time‐Inverted Lagrangian Transport (STILT) model. *Journal of Geophysical Research: Atmospheres*, 108(D16), https://doi.org/10.1029/2002JD003161. 2003. 

**STILT error uncertainties**: 
Lin, J. C., and Gerbig, C., Accounting for the effect of transport errors on tracer inversions, *Geophys. Res. Lett.*, 32, L01802, https://doi.org/10.1029/2004GL021127, 2005.

## Model applications 
**X-STILT for TROPOMI XCO**: Wu, D., Liu, J., Wennberg, P. O., Palmer, P. I., Nelson, R. R., Kiel, M., and Eldering, A.: Towards sector-based attribution using intra-city variations in satellite-based emission ratios between CO2 and CO, Atmos. Chem. Phys. Discuss. [preprint], https://doi.org/10.5194/acp-2021-1029, in review, 2022.



HISTORY.md
Commit by June 19, 2023
============
Code available for simulating tropospheric NO2 columns!


Commit by June 16, 2022
============
Code refactoring for transport error estimates and sounding selections for non-CO2 species. 

:pushpin: Sounding selections:

* `nf_dlat` and `nf_dlon`: define the NEAR-FIELD regions around `site`
* `dlat` and `dlon`: define the ENTIRE target regions around `site`

* `num_nf_lat` * `num_nf_lon`: total # of soundings to be selected along either the latitudinal or longitudinal direction (e.g., ~7 * 14 total soundings over the NEAR-FIELD domain around `site`)

* `num_bg_lat` * `num_bg_lon`: total # of soundings to be selected along either the latitudinal or longitudinal direction (e.g., ~3 * 3 total soundings over the FAR-FIELD or BACKGROUND domain around `site`)

* `get.recp.sensorv2()`: function for the selection of satellite soundings to save computational cost; If selected near- or far-field soundings share the same coordinate, only one of them will be treated as model receptor. 

<img src="figs/demo_selection.png" width="400" height="400">

Figure - An example of sounding selections (though not highly recommended for TROPOMI given its coarse resolution). 


Commit by July 6, 2021
============
Changes to ['merge_stilt-hysplit' branch](https://github.com/uataq/X-STILT/tree/merge_stilt-hysplit) towards the release of X-STILT version 2 that can work with non-OCO sensors or without the dependence of satellite data:

:pushpin: X-STILT can now work with TROPOMI X-gas data, such as 
   * footprint weighting using TROPOMI species-specific averaging kernels;

   * functions to grab vertical column density and convert to vertical mixing ratio in ppb; 

   * need to specify TROPOMI info via `obs_sensor` and `obs_species` that are currently limited to CO, CH4, and NO2;

:pushpin: IDEAL simulations for column footprints/enhancements from any locations (no requirement of satellite data).
   * To use the ideal mode, `obs_sensor` needs to be set as NA; 
   * Column AK is simply treated as 1 from TOA to the surface (`ak_wgt` is forced as FALSE);
   * Column PWF is calculated using modeled variables/profiles (`pwf_wgt` = TRUE). 

   * Must provide a .csv/.txt file containing the latitudes (`lati`) and longitudes (`long`) of your desired receptor locations. Receptor time (in form of YYYYMMDDHH or YYYYMMDDHHss) can either be included in the csv/txt file as the 3rd column (with name of `time`) OR assigned to `timestr` in `run_xstilt.r`. See example of `receptor_demo.csv`. 
   
   * [USE WITH CAUTIONS] NOT suitable for comparisons with real satellite sensors that have variable vertical sensitivities.

:pushpin: **Built upon the latest HYSPLITv5**: Replace discrete release levels with *line source release* -- i.e., particles are now evenly distributed between `minagl` and `maxagl`, e.g., default is 3000 particles from 0 - 3km; 

:pushpin: PWF calculations in `r/src/recp_trajec_foot`

   * Instead of interpolate PWF from OCO-2 to each discrete release level :arrow_forward: calculate PWF based on dp and dry air column density (mol m-2) between particles. 

   * To calculate the dry air X-density, we extracted specific humidity and temperature profiles via `get.met.vars()`. 

   * If ak.wgt is turned off (`ak.wgt` == FALSE), the dependence on any satellite sensors will be removed. That being said, normalied AK is 1 and PWF is calculated based on MODELED surface pressures instead of RETRIEVED surface pressures. 

:pushpin: Other minor changes: 
   * Allow the calculation of another set of footprint with a different temporal resolution from the default run at the same time, e.g., hourly footprint implied by `time_integrate2 = F`;

   * Add interactive mode for choosing site, overpass time, etc. One can modify the code to turn it off;

   * Insert the google API as the first line of `insert_ggAPI.csv` that will be sourced by the model for geolocation and google map plotting; 

   * Get rid of the additional .rds files end with X_wgttraj.rds to reduce file sizes :arrow_forward: trajectory-level footprints after the AK and PWF weighting are now stored as additional columns in the initial *X_traj.rds files. Here are the descriptions of columns in output$particle from rds files: 
      * `foot_before_weight:` footprint BEFORE AK and PWF weighting;
      * `foot:` footprint AFTER AK and PWF weighting; 
      * `pwf, ak.norm, ap, ak.pwf:` computed pressure weighting and interpolated apriori and averaging kernel profiles from OCO-2/3; 
      * `wgt:` final weighting factors used for scaling trajectory-level footprint based on flags of `ak.wgt` and `pwf.wgt`. 


run_xstilt.r
#' create a namelist for running X-STILT trajectories
#' @author Dien Wu
#'
# ---------------------------------------------------------------------------
#' A. run_trajec == T or run_foot == T: run trajectory or footprint, simulations
#' A1. if run_hor_err == T or run_ver_err == T 
#'     X-STILT generates trajec with perturbations from wind or PBL errors.
#'     along with traj-level CO2 and its error statistics (see AA1 for stats)
# -----------------------------------------------------------------------------
#' AA1. if run_hor_err == T & run_wind_err == T
#'      X-STILT calculates wind uncertainties and biases using NOAA radiosonde
#'      data, one needs to download FSL data from https://ruc.noaa.gov/raobs/
#'      and put data under @param raob_path as @param raob_fn
# -----------------------------------------------------------------------------
#' A2. if run_emiss_err == T 
#'     X-STILT generates trajec and footprint with a resolution of 0.1deg, 
#'     since the absolute emission error has res of 0.1deg. 
#' -----------------------------------------------------------------------------
#' B. run_trajec == F & run_foot == F & run_sim = T: requires trajec and foot
#'    X-STILT models XCO2 enhancements based on FFCO2 emission from 1km ODIAC;
#'                or XCO2 errors due to emission errors (run_emiss_err == T);
#'                or XCO2 errors due to transport errors (run_hor_err == T).
#' ---------------------------------------------------------------------------
#' Instructions for ideal column simulations 
#' *** AK is simply treated as 1 (@param ak_wgt == FALSE)
#' *** PW is calculated using modeled met variables (@param pwf_wgt == TRUE)
#' *** The user NEEDS to provide a txt or csv file for receptor locations
#' *** correct file should include long, lat, and time (optional, YYYYMMDDHH). 
#' *** You can also assign your receptor time to @param timestr. HOWEVER it'll 
#' *** overwrite the existing time column in the receptor file if there is. 
# ---------------------------------------------------------------------------
# 
# Latest major changes, DW
# 1. upgrade to STILT-HYSPLITv5
# 2. allow for ideal sim w/o obs using a csv file provided by users, 12/06/2020
# 3. allow for working with TROPOMI data, 07/03/2021
# 4. allow for sounding selection (OCO and TROPOMI) within NEAR- or FAR-FIELD, 
#    see comments around L53-60, L154-160), 06/17/2022
# ---------------------------------------------------------------------------

# !!!! to run this code in R session: source('run_xstilt.r')
options(timeout = max(1500, getOption('timeout')))   # for download.file

# ----------- Dependencies and API key for geolocation (must-have) ---------- #
## go to X-STILT dir and source functions and libraries
homedir = '/central/home/dienwu'
xstilt_wd = file.path(homedir, 'X-STILT') 
setwd(xstilt_wd); source('r/dependencies.r')

# Please insert your API in the 'insert_ggAPI.csv' for use of ggplot and ggmap
# google API can be obtained from https://console.developers.google.com/
api.key = readLines('insert_ggAPI.csv')

# --------------------- Site location params (must-have) --------------------- #
cat('Enter your site name: ')
site = readLines('stdin', n = 1); print(site)

# define entire- and near-field domain (usually for high observations)
# for selecting observations, if num_* is not NA
# ENTIRE domain - site_lat +/- dlat, site_lon +/- dlon 
# NEAR-FIELD domain - site_lat +/- nf_dlat, site_lon +/- nf_dlon
dlon = 0.5            # e.g., dlon = 0.5 means 1 x 1 degree box around the site
dlat = 0.5            # dlat/dlon in degrees 
nf_dlon = 0.3
nf_dlat = 0.3

# - automatically obtain lat/lon (requires Google API key in insert_ggAPI.csv)
# - OR mannually insert coordinates, e.g., site.loc = data.frame(lon = , lat = )
lon_lat = get.lon.lat(site, dlon, dlat, site.loc = NULL)


# ------------------------ I/O params (must-have) --------------------------- #
# as a default, all input data are stored under input_path, change if you like
input_path  = '/central/groups/POW'
store_path  = file.path(input_path, 'XSTILT_output', site)

# *** modify the info path/directory that stores OCO-2/3 or TROPOMI data -------
obs_sensor  = c('OCO-2', 'OCO-3', 'TROPOMI', NA)[3]
obs_ver     = c('V11r', 'V10p4r', NA)[3]       # retrieval algo ver if there is
obs_species = c('CO2', 'CO', 'NO2', 'CH4')[3]  # only 1 gas per run
oco_path = file.path(input_path, obs_sensor, paste0('L2_Lite_FP_', obs_ver))
sif_path = file.path(input_path, obs_sensor, paste0('L2_Lite_SIF_', obs_ver))
trp_path = file.path(input_path, obs_sensor, obs_species, 'L2')
store_path = file.path(input_path, 'XSTILT_output', site)
if (!is.na(obs_ver)) store_path = file.path(store_path, obs_ver)


# *** if obs_sensor is NA, perform ideal runs without satellite dependence, 
# *** instead, the user needs to provide a .csv/.txt file for receptor info
# It should contain desired longitude ('long') and latitude ('lati'), 
# one can also include a third column for receptor time ('time') OR enter the 
# receptor time to `timestr` below. Each column needs to be separated by comma. 
if (is.na(obs_sensor)) {   # X-simulations WOUT satellite data
  recp_fn = 'receptor_demo.csv'                               # <path/filename>
  obs_ver = obs_species = obs_path = sif_path = NA

} else {                   # X-simulations WITH satellite data
  recp_fn = NA
  obs_path = ifelse(obs_sensor == 'TROPOMI', trp_path, oco_path)
}

# paths for radiosonde, ODIAC, chemical transport model (optional) -------------
raob_path  = file.path(input_path, 'RAOB', site)  # NOAA radiosonde
raob_fn    = list.files(raob_path, 'tmp')
odiac_ver  = c('2019', '2020')[2]         # ODIAC version
odiac_path = file.path(input_path, 'ODIAC', paste0('ODIAC', odiac_ver))  
cat('Done with choosing cities & overpasses...\n')
# --------------------------------------------------------------------------- #


# ---------------------- obtaining overpass time string --------------------- #
#' @param obs_sensor == TROPOMI - daily obs, time string provided by user
#' @param obs_sensor == NA - ideal run w/o obs, time string provided by user
#' @param timestr in format of YYYYMMDD or YYYYMMDDHH (either works)
timestr = '20191007'

#' @param obs_sensor == OCO-2/3 - can help search for overpasses with #
#' sufficient data over the entire area and/or near-field area around site
if (grepl('OCO', obs_sensor)) 
  timestr = get_timestr(site, lon_lat, obs_sensor, obs_ver, obs_path, 
                        store_path, recp_fn, plotTF = FALSE, qfTF = TRUE, 
                        nfTF = TRUE, nf_dlon, nf_dlat, sif_path, 
                        searchTF = FALSE, date_range = c('20140101','20221231'))

cat('Done with choosing cities & overpasses...\n')
# --------------------------------------------------------------------------- #


# --------------------- Basis X-STILT flags (must-have) --------------------- #
run_trajec    = T    # if to generate trajec; T: may overwrite existing trajec
run_foot      = T    # if to generate footprint
run_hor_err   = F    # T: set error parameters
run_ver_err   = F    # T: set error parameters
run_emiss_err = F    # T: get XCO2 error due to prior emiss err
run_wind_err  = F    # T: calc wind error based on RAOB 
run_sim       = F    # T: calc XFF or error with existing foot (only for CO2)
nhrs          = -12  # number of hours backward (-) or forward (+)

# output variable names required in trajec.rds
varstrajec = c('time', 'indx', 'lati', 'long', 'zagl', 'zsfc', 'foot', 'samt',
               'dmas', 'mlht', 'temz', 'pres', 'sigw', 'tlgr', 'dens', 'sphu')


# ---------------------- Receptor params (must-have) ------------------------ #
# line source for column release according to HYSPLITv5, DW, 09/17/2020
# all particles are roughly evenly distributed between minagl and maxagl
minagl = 0                # min release height in meter AGL
maxagl = 3000             # max release height in meter AGL
numpar = 3000             # total number of particles between minagl and maxagl

# Receptor selection (params are set as NA for ideal sim) ---------------------
# T: create receptors within each satellite sounding besides centered lats/lons
# F: place receptors ONLY at the centered lat/lon of a sounding, DW, 07/02/2021
jitterTF   = FALSE            # T: works better for TROPOMI with larger polygons
num_jitter = 5                # number of additional receptors per sounding

#' Only place receptors for soundings that qualify @param obs_filter
#' here are some choices for OCO-2/3 and TROPOMI (uncomment the one you need)
#obs_filter = c('QF', 0)             # select OCO soundings with QF = 0
obs_filter = c('QA', 0.7)            # select TROPOMI soundings with QA >= 0.5 
#obs_filter = NULL                   # use all soundings regardless of QF or QA

#' evenly select soundings within near-field or far-field (background) regions
#' near-field: defined by @param nf_dlat & @param nf_dlon 
#' far-field: defined by @param dlat & @param dlon except for near-field region
# num of soundings/receptors along lat or lon within farfield or nearfield
# if ALL are NA - NO need to select soundings (see github README for more clues)
num_bg_lat = num_bg_lon = num_nf_lat = num_nf_lon = NA                
#num_bg_lat = 5; num_bg_lon = 5; num_nf_lat = 10; num_nf_lon = 10


# ------------------- ARL format meteo params (must-have) -------------------- #
# see STILTv2 https://uataq.github.io/stilt/#/configuration?id=meteorological-data-input
met = c('gfs0p25', 'hrrr', 'wrf27km')[1]         # choose met fields
met_path = file.path(homedir, met)               # path of met fields
met_file_format = '%Y%m%d'                       # met file name convention
n_met_min = download.met.arl(timestr, met_file_format, nhrs, met_path, met)

# OPTION for subseting met fields if met_subgrid_enable is on, 
# useful for large met fields like GFS or HRRR
met_subgrid_buffer = 0.1   # Percent to extend footprint area for met subdomain
met_subgrid_enable = F   

# if set, extracts the defined number of vertical levels from the meteorological 
# data files to further accelerate simulations, default is NA
met_subgrid_levels = NA    
cat('Done with params for receptor and meteo- setup...\n')


# -------------------- Footprint params (must-have) ------------------------- #
# whether weight footprint by AK and PW for column simulations 
# if F, AK = 1; if T, look for AK at the closest sounding from real data
ak_wgt  = TRUE         # *** if obs_sensor is NA, ak_wgt is forced as FALSE 
pwf_wgt = TRUE
if (is.na(obs_sensor)) ak_wgt = FALSE 

# footprint spatial domain defined as site.lat +/- foot_dlat and 
# site.lon +/- foot_dlon in degrees, 10 here meaning 20 x 20deg box around site
foot_dlat = 10     
foot_dlon = 10 

# ---------------------------------------------------------------------------- #
# *** You can run multiple sets of footprints with different spatiotemporal 
#     resolution from the same trajec at one time (see below), DW, 07/01/2021
#' @param MAIN_RUN (e.g., 1km time-integrated footprint) -----------------------
foot_res  = 1/10       # spatial resolution in degree
foot_nhrs = nhrs        # if foot_nhrs < nhrs, subset trajec before footprint
time_integrate = TRUE   # F: hourly foot; T: time-integrated foot

#' @param OPTIONAL_RUN with diff configurations (e.g., hourly 0.05 deg foot) ---
#' both params can be a vector, footprint filename contains res info
#' if no need for alternative runs, set @param foot_res to NA, DW, 02/11/2019
foot_res2 = c(NA, 1/10, 1/20, 1)[1]    # spatial res in degree
time_integrate2 = FALSE                # ndim(time_integrate2) = ndim(foot_res2)

# ---------------------------------------------------------------------------- #
# other neccesary footprint params using STILTv2 (Fasoli et al., 2018)
hnf_plume     = TRUE                # T: hyper near-field (HNP) for mixing hgts
smooth_factor = 1                   # Gaussian smooth factor, 0 to disable
projection    = '+proj=longlat'
cat('Done with params for error analysis and footprint setup...\n')
# ---------------------------------------------------------------------------- #


# ------- Transport error params (only needed for CO2 error, OPTIONAL) ------- #
# if run_hor_err = T, require ODIAC and CT fluxes and mole fractions
# to calculate horizontal transport error of total CO2, DW, 07/28/2018
if (run_hor_err) {
  ct_ver      = ifelse(substr(timestr, 1, 4) >= '2016', 'v2017', 'v2016-1')
  ct_path     = file.path(input_path, 'CT-NRT', ct_ver)
  ctflux_path = file.path(ct_path, 'fluxes/optimized')
  ctmole_path = file.path(ct_path, 'molefractions/co2_total')
} else ct_ver = ctflux_path = ctmole_path = NA       # end if run_hor_err

# if run_ver_err = T, calculate vertical transport error
# if run_ver_err = F, set zisf = 1 (default)
zisf = c(0.6, 0.8, 1.0, 1.2, 1.4)[3]; if (!run_ver_err) zisf = 1.0

# if run_emiss_err = T, calculating XCO2 error due to emiss error, 
# need EDGAR and FFDAS files to compute emission spread, DW, 10/21/2018
if (run_emiss_err) { 
  edgar_file = file.path(input_path, 'EDGAR/v42_CO2_2008_TOT.0.1x0.1.nc')
  ffdas_path = file.path(input_path, 'FFDAS')
  ffdas_file = list.files(ffdas_path, 'totals')
  ffdas_file = file.path(ffdas_path, ffdas_file[grep('2008', ffdas_file)])
} else edgar_file = ffdas_file = NA                   # end if run_emiss_err


# ----------------------------- SLURM params -------------------------------- #
# avoid using too many e.g., > 10 cores per node -> memory limits
# set slurm to FALSE, if system does not support slurm
# *** there is a current bug with the rslurm package, you may need to 
# mannually pull the development version from github by doing:
#devtools::install_github('SESYNC-ci/rslurm')               # DW, 11/6/2020
slurm = T                           # T: SLURM parallel computing
n_nodes = 12
n_cores = 10
if (!slurm) n_nodes = n_cores = 1
timeout = 12 * 60 * 60              # time allowed before terminations in sec
job_time = '12:00:00'               # total job time
slurm_account = 'pow'
slurm_partition = 'any'

# *** IF YOU EVER RAN INTO OOM-KILL ERROR, YOU CAN ENLARGE THE MAX MEM HERE *** 
# The ammount of memory per node you need in MB, extending to 10 GB per core
# given 180+ GB max memory on each Caltech cluster with 32 cores
# **** PLEASE adjust mem_per_* based on your machine
mem_per_core = 10                                 # max memory per core in GB
mem_per_node = n_cores * mem_per_core * 1024      # max mem per node now in MB 


# namelist required for generating trajec
namelist = list(ak_wgt = ak_wgt, ct_ver = ct_ver, ctflux_path = ctflux_path, 
                ctmole_path = ctmole_path, edgar_file = edgar_file, 
                ffdas_file = ffdas_file, foot_res = foot_res, 
                foot_res2 = list(foot_res2), foot_nhrs = foot_nhrs, 
                foot_dlat = foot_dlat, foot_dlon = foot_dlon, 
                hnf_plume = hnf_plume, jitterTF = jitterTF, 
                job_time = job_time, lon_lat = list(lon_lat), 
                mem_per_node = mem_per_node, met = met,                 
                met_file_format = met_file_format, met_path = met_path, 
                met_subgrid_buffer = met_subgrid_buffer, 
                met_subgrid_enable = met_subgrid_enable, 
                met_subgrid_levels = met_subgrid_levels, minagl = minagl, 
                maxagl = maxagl, nhrs = nhrs, n_cores = n_cores, 
                n_met_min = n_met_min, n_nodes = n_nodes, nf_dlat = nf_dlat, 
                nf_dlon = nf_dlon, num_jitter = num_jitter, 
                num_bg_lat = num_bg_lat, num_bg_lon = num_bg_lon, 
                num_nf_lat = num_nf_lat, num_nf_lon = num_nf_lon, 
                numpar = numpar, obs_filter = list(obs_filter),
                obs_path = obs_path, obs_sensor = obs_sensor, 
                obs_species = obs_species, odiac_path = odiac_path, 
                odiac_ver = odiac_ver, projection = projection, 
                pwf_wgt = pwf_wgt, raob_fn = raob_fn, recp_fn = recp_fn, 
                run_emiss_err = run_emiss_err, run_foot = run_foot, 
                run_hor_err = run_hor_err, run_sim = run_sim, 
                run_trajec = run_trajec, run_ver_err = run_ver_err, 
                run_wind_err = run_wind_err, site = site, slurm = slurm, 
                slurm_account = slurm_account,slurm_partition = slurm_partition,
                smooth_factor = smooth_factor, store_path = store_path, 
                time_integrate = time_integrate, 
                time_integrate2 = list(time_integrate2), 
                timeout = timeout, timestr = timestr, varstrajec = varstrajec, 
                xstilt_wd = xstilt_wd, zisf = zisf)
cat('Done with creating namelist...\n')

config_xstilt(namelist)  # start X-STILT, either calc traj, foot or simulation
q('no')
# end of main script
# ---------------------------------------------------------------------------- #